### PR TITLE
mmap: check offset and size more carefully

### PR DESCRIPTION
**Pull request by yuwata**
Also fixes several issues found by coverity.

### CODE REVIEW

File: src/libsystemd/sd-journal/mmap-cache.c [hunk 1]
```diff
@@ -278,41 +277,49 @@ static int add_mmap(
         assert(size > 0);
         assert(ret);
 
-        woffset = offset & ~((uint64_t) page_size() - 1ULL);
-        wsize = size + (offset - woffset);
-        wsize = PAGE_ALIGN(wsize);
+        /* PAGE_ALIGN() and friends used below take size_t. If we pass too large value these may provide
+         * an unexpected result. */
+        if (offset > (uint64_t) SIZE_MAX)
```
Comment (daandemeyer) [inline src/libsystemd/sd-journal/mmap-cache.c]: Shouldn't we then fix the macros so they can work with `size_t`? It makes sense to me that we'd support journals larger than 4Gib even on 32-bit systems.

Comment (yuwata) [reply src/libsystemd/sd-journal/mmap-cache.c]: OK. Pushed another version. PTAL.

File: src/basic/macro.h [hunk 2]
```diff
@@ -42,6 +42,8 @@
 /* test harness */
 #define EXIT_TEST_SKIP 77
 
+#define OFF_T_MAX ((off_t) -1)
```
Comment (poettering) [inline src/basic/macro.h]: off_t is actually signed, hence this value is not actually a maximum at all

File: src/basic/memory-util.h [hunk 3]
```diff
@@ -12,9 +12,12 @@
 #include "memory-util-fundamental.h"
 
 size_t page_size(void) _pure_;
-#define PAGE_ALIGN(l) ALIGN_TO((l), page_size())
-#define PAGE_ALIGN_DOWN(l) ((l) & ~(page_size() - 1))
-#define PAGE_OFFSET(l) ((l) & (page_size() - 1))
+#define PAGE_ALIGN(l)         ALIGN_TO(l, page_size())
+#define PAGE_ALIGN_64(l)      ALIGN_TO_64(l, page_size())
```
Comment (poettering) [inline src/basic/memory-util.h]: so looking at this one might get the idea this was aligning to 64bit multiples. Which it of course is not. But if we rename this to PAGE_ALIGN_U64() then I think everything is much clearer, as everyone knows that u64 refers to a unsigned in widely.

File: src/libsystemd/sd-journal/journal-file.c [hunk 4]
```diff
@@ -47,30 +47,31 @@
 #define DEFAULT_COMPRESS_THRESHOLD (512ULL)
 #define MIN_COMPRESS_THRESHOLD (8ULL)
 
+#define SIZE64_KB UINT64_C(1024)
```
Comment (poettering) [inline src/libsystemd/sd-journal/journal-file.c]: maybe U64_KB?

**Comment by yuwata** [comment]
@poettering Thank you for the review. All comments are addressed. PTAL.

**Review by bluca — APPROVED** [review]

**Pull request merged**